### PR TITLE
WIP: Coverage for cython generated files

### PR DESCRIPTION
**Pull request by bashtage**
Working example with coverage. pyx files have coverage.  Big run time increase but manageable.

xref #4020

### CODE REVIEW

**Comment by bashtage** [comment]
This isn't for merging.

**Comment by josef-pkt** [comment]
This pull request introduces 4 alerts - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-279f5c24433f4e41daa604ee5edd3a34469954b7)

**new alerts:**

* 1 for Comparison using is when operands support __eq__
* 1 for Unreachable code
* 1 for Suspicious unused loop iteration variable
* 1 for Variable defined multiple times

---

*Comment posted by [lgtm.com](https://lgtm.com)*

**Comment by josef-pkt** [comment]
This pull request introduces 2 alerts - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-15b28d2014ceb9e92ad94dccedc16c148cf3ef22)

**new alerts:**

* 1 for Unreachable code
* 1 for Suspicious unused loop iteration variable

---

*Comment posted by [lgtm.com](https://lgtm.com)*

**Comment by josef-pkt** [comment]
This pull request introduces 2 alerts - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-72c790c7ac5cd126143c25a7e150173ac90d2a59)

**new alerts:**

* 1 for Unreachable code
* 1 for Suspicious unused loop iteration variable

---

*Comment posted by [lgtm.com](https://lgtm.com)*

**Comment by bashtage** [comment]
@jbrockmendel This works fine now.  

This should work:

```bash
git clean -xfd
export CYTHON_COVERAGE=1  # Not really needed, setup_coverage defaults to coverage
python setup_coverage.py develop
pytest --cov=statsmodels --cov-config=.coveragerc --cov-report=html statsmodels
```

**Comment by josef-pkt** [comment]
This pull request introduces 2 alerts - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-b7b74b251def35bf1e0eaa189b86a98fee13ec9e)

**new alerts:**

* 1 for Unreachable code
* 1 for Suspicious unused loop iteration variable

---

*Comment posted by [lgtm.com](https://lgtm.com)*

**Comment by josef-pkt** [comment]
This pull request introduces 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-9aa4651b19b6083311243111d5ce74e1a39106c2)

**new alerts:**

* 1 for Suspicious unused loop iteration variable

---

*Comment posted by [lgtm.com](https://lgtm.com)*

**Comment by bashtage** [comment]
Probably impossible to run cython coverage on Travis since times increase substantially.  When running only the cythonized modules, test times goes from ~4 minutes to ~25.

Here is a report showing coverage of pyx files

https://coveralls.io/jobs/30323075

Look under All, then filter by pyx.  Can only see the line by line for the pyx that are in the repo. It isn't possible to see the line-by-line coverage for the generated files.

**Comment by josef-pkt** [comment]
This pull request introduces 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-8bfb6f5bfbcb77bd023d713ca474c8d0ec80089d)

**new alerts:**

* 1 for Suspicious unused loop iteration variable

---

*Comment posted by [lgtm.com](https://lgtm.com)*

**Comment by jbrockmendel** [comment]
> Here is a report showing coverage of pyx files

This is great --  the fact that this is working, not so much the results.  Thanks for getting this over the finish line.

The coverage for the statespace files is lower than I expected -- Chad put together an impressive test suite.  @ChadFulton any thoughts?

Getting coverage for the pyx files generated from pyx.in files would be helpful.  I'll take a swing at that locally.

**Comment by bashtage** [comment]
Most of the low coverage is due to the 4x repeat in all files for (s,d,c,z)
types.  A file that is 25% covered probably has one path checked (usually
d).  It might be worth considering coverage for one of the complex cases
since there is an increased chance of some strange result.  If d works, I
would expect s to work unless some helper function is missing (unlikely).

**Comment by bashtage** [comment]
Have substantiallly rewritten setup to setup_coverage in this PR, I think it would be helpful to modernize it.  I think this is a good starting point and only needs some work on a `build_ext` to defer importing of NumPy (so that it could be installed in an empty virtualenv) and Cython (since this isn't required). These can probably be mostly cribbed from pandas. 

I also added versioneer, which I have really liked using since it makes versioning so trivial. 

One thing I also did was to move the "user" variables to the top so that they are easy to find since they might change over time (e.g., requirements).

**Comment by josef-pkt** [comment]
If we want to go the route of using a separate branch, e.g. master_coverage, for the CI testing, then we could commit the generated pyx, instead of git-ignoring them, so that they are available in coveralls.

**Comment by bashtage** [comment]
I don't think a separate branch is a good idea.  It will be really hard to keep in sync.  I think the best idea would be to have an improved setup that handles the cython coverage when requested (e.g. when `CYTHON_COVERAGE` is set) but otherwise works just like the usual setup.  

Online coverage reports aren't that useful for Cython code -- anyone wants to work on this or check their code (if there was a new Cython file) could just do so locally with a HTML report.

**Comment by josef-pkt** [comment]
This pull request introduces 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-2fd712fc0d557aece0acd86f8be10b1ddd4cdf2f)

**new alerts:**

* 1 for Suspicious unused loop iteration variable

---

*Comment posted by [lgtm.com](https://lgtm.com)*

File: .coveragerc [hunk 1]
```diff
@@ -1,28 +1,38 @@
...
+    */tests/*
+    */datasets/*
+
 # Regexes for lines to exclude from consideration
 exclude_lines =
     # Have to re-enable the standard pragma
     pragma: no cover
 
+    # Don't complain if tests don't hit defensive assertion code:
+    raise NotImplementedError
```
Comment (jbrockmendel) [inline .coveragerc]: Not necessarily for this Pr, but I think we shouldn’t gnome NotImplmentedError

Comment (bashtage) [reply .coveragerc]: The legit reason to not count these is that Abstract classes often use NIE to force subclasses to implement methods.

Comment (jbrockmendel) [reply .coveragerc]: That’s reasonable, but in practice in sm it usually means someone stopped writing a function halfway through.

One option would be to use pandas AbstractMethodError for the use case you mentioned.

File: .coveragerc [hunk 2]
```diff
@@ -17,12 +23,16 @@ exclude_lines =
...
     # Don't complain if non-runnable code isn't run:
     if 0:
     if __name__ == .__main__.:
 
-ignore_errors = False
+include = */statsmodels/*
+ignore_errors = True
 
 [html]
 directory = coverage_html_report
```
Comment (jbrockmendel) [inline .coveragerc]: GH suggests a \n is missing.

File: tox.ini [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    OPENBLAS_NUM_THREADS=1
+changedir = {envdir}
+
+[testenv:coverage]
+setenv=
+    {[testenv]setenv}
+    CYTHON_COVERAGE=1
+commands=
+    coverage erase
+    python setup.py develop
```
Comment (jbrockmendel) [inline tox.ini]: build_ext?

**Comment by bashtage** [comment]
Closing in favor of #5188

**Pull request closed without merging**